{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47b89519-b51b-4313-a602-28440c16b672",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations\n",
    "from scipy.stats import zscore\n",
    "from IPython import display as disp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a8bf51d-d5c6-4555-baff-944c69d6c018",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfolder = \"/Users/fekeann/Documents/Lab/RNAseq/renormalized_data/correlations/newSynt/\"\n",
    "if not os.path.exists(outfolder):\n",
    "    os.mkdir(outfolder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27075260-f9cf-4f10-bba8-c559b223299b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_simple = [\"ClosedFlower\", \"HookedStolon\", \"ImmFruit\", \"MatureFruit\",\n",
    "                 \"OpenFlower\", \"Sprout\", \"StemControl\", \"SwollenStolon\",\n",
    "                 \"TuberS1\", \"TuberS2\", \"TuberS3\", \"TuberS4\", \"TuberS5\",\n",
    "                 \"YL10am\", \"ColdLeafControl\", \"RootControl\"] \n",
    "                    # \"ColdLeafControl was chosen at random as a leaf control, \n",
    "                    # since I believe its a more mature leaf than \"YL\"\n",
    "                    # all/most samples were taken at 10am so that's why I picked YL10am\n",
    "\n",
    "stress_with_control = [\"BTH\", \"BTHControl\", \"ColdLeaf\", \"ColdLeafControl\",\n",
    "                       \"DroughtLeaf\", \"DroughtRoot\", \"Heat\", \"HeatControl\",\n",
    "                       \"Meja\", \"MejaControl\", \"RootControl\", \"SaltLeaf\",\n",
    "                       \"SaltLeafControl\", \"SaltRoot\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4d01563-7b0b-4e27-85d5-aa0c2030bd63",
   "metadata": {},
   "outputs": [],
   "source": [
    "shd_trace_file = \"/Users/fekeann/Documents/Lab/RNAseq/renormalized_data/rlogs/trace_format/full_diel_norm_atl_shd_leaf.csv\"\n",
    "shd_tuber_trace_file = \"/Users/fekeann/Documents/Lab/RNAseq/renormalized_data/rlogs/trace_format/full_diel_norm_atl_shd_tuber.csv\"\n",
    "lgd_trace_file = \"/Users/fekeann/Documents/Lab/RNAseq/renormalized_data/rlogs/trace_format/full_diel_norm_atl_lgd_leaf.csv\"\n",
    "tissue_rlog_all = \"/Users/fekeann/Documents/Lab/RNAseq/Tissues/deseq_outputs/all_tissues_outfile.csv\"\n",
    "synt_file = '/Users/fekeann/Documents/Lab/RNAseq/Syntelogs/final_selections/publication_output/syntelogs_atl_only.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e0ff908-7d9e-4921-9d40-34582c631ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "tc_dict = {\"shd\":shd_trace_file,\n",
    "           \"lgd\":lgd_trace_file,\n",
    "           \"tuber\":shd_tuber_trace_file,\n",
    "           \"tissue\":tissue_rlog_all}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "192f5a9b-034d-4373-bbe0-9ff0c955d829",
   "metadata": {},
   "outputs": [],
   "source": [
    "syntelogs = pd.read_csv(synt_file)\n",
    "syntelogs = syntelogs[syntelogs.geneID.str.contains(\"Atl\")].copy()\n",
    "syntelogs[\"Haplotype\"] = syntelogs.geneID.apply(lambda x: x.split(\"_\")[-1][0])\n",
    "\n",
    "#get rid of \"S\" haplotypes; we only want those with known chromosomes\n",
    "syntelogs = syntelogs[syntelogs.Haplotype.isin([\"0\", \"1\",\"2\",\"3\",\"4\"])].copy()\n",
    "\n",
    "# remove isoform information from the geneIDs\n",
    "syntelogs[\"geneID\"] = syntelogs.geneID.apply(lambda x: \".\".join(x.split(\".\")[:-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9b64486-1726-4b8d-b344-6e1a783088f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def avg_zscore_format(path, dtype=\"timecourse\", use_cols = None):\n",
    "    \"\"\"\n",
    "    Converts trace files with replication to the z-scored version, averaged by time\n",
    "    INPUTS:\n",
    "        path:     str, the filepath to the rlog trace file\n",
    "        dtype:    str, refers to what type of dataset is found in path\n",
    "                  options: \"timecourse\" or \"tissue\"; \n",
    "                  \"timecourse\" is used for LgD or ShD (leaf or tuber)\n",
    "                  \"tissue\" is used for stress or tissue datasets\n",
    "        use_cols: str or None, refers to which columns to use for \"tissue\" dtype\n",
    "                  options: \"tissue\" or \"stress\"\n",
    "                  references the lists tissue_simple or stress_with_control in cell 3\n",
    "\n",
    "    OUTPUTS:\n",
    "        A pandas df containing the z-scored, time-averaged traces\n",
    "    \"\"\"\n",
    "    \n",
    "    data = pd.read_csv(path)\n",
    "\n",
    "    # the tissue file doesn't have \"geneID\" as the column label\n",
    "    if dtype == \"tissue\":\n",
    "        data.rename({\"Unnamed: 0\":\"geneID\"}, axis=1, inplace=True)\n",
    "    \n",
    "    data.set_index(\"geneID\", inplace=True)\n",
    "    \n",
    "    if dtype == \"tissue\":\n",
    "        # we only care about the first two terms of the header; \n",
    "        # the replicate information is required since we can't have two identically named columns\n",
    "        data.rename({col: \"_\".join(col.split(\"_\")[:2]) for col in data.columns},axis=1, inplace=True)\n",
    "        tissue_dict = {}\n",
    "        # filter the dataset to just having the appropriate columns for the analysis being done\n",
    "        for col in data.columns:\n",
    "            sample = col.split(\"_\")[0]\n",
    "            tissue_dict[col] = sample\n",
    "        if use_cols == \"stress\":\n",
    "            data = data[[col for col in data.columns if tissue_dict[col] in stress_with_control]]\n",
    "        elif use_cols == \"tissue\":\n",
    "            data = data[[col for col in data.columns if tissue_dict[col] in tissue_simple]]\n",
    "\n",
    "    #convert from wide to long; sample type is now \"time\" even if it's a tissue/stress.\n",
    "    data = data.stack().reset_index().rename({\"level_1\":\"time\", 0:\"mean_expression\"}, axis=1)\n",
    "\n",
    "    #get rid of replicate number from the sample nfo\n",
    "    if dtype == \"tissue\":\n",
    "        data[\"time\"] = data[\"time\"].apply(lambda x: x.split(\"_\")[0])\n",
    "    else:\n",
    "        data[\"time\"] = data[\"time\"].apply(lambda x: int(x.split(\"_\")[0].split(\"T\")[-1]))\n",
    "        \n",
    "    #average the expresson within gene and condition\n",
    "    data = data.groupby(['geneID', 'time']).mean().reset_index().set_index(\"geneID\")\n",
    "\n",
    "    # calculate the z-score for the averaged dataset\n",
    "    zscores = data.pivot(columns=\"time\").apply(lambda row: zscore(row, ddof=1), axis=1)\n",
    "    zscores = zscores.droplevel(level=0, axis=1)\n",
    "\n",
    "    #remove isoform information from geneID\n",
    "    zscores[\"geneID\"] = zscores.index.map(lambda x: \".\".join(x.split(\".\")[:-1]))\n",
    "    zscores.set_index(\"geneID\", inplace=True)\n",
    "\n",
    "    return zscores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ed8cd1f-4445-4eb0-a679-8f65620125a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_pair_corr(zscores, pair):\n",
    "    \"\"\"\n",
    "    Calculate the correlation of two expression patterns; returning None if one is un-expressed\n",
    "    INPUTS:\n",
    "        zscores: a df containing the z-scored expression patterns, indexed by \"geneID\"\n",
    "        pair:    a list of strs of length 2, containing the geneID of the genes to be correlated\n",
    "\n",
    "    OUTPUT:\n",
    "        a float or None, representing the correlation between the two patterns\n",
    "    \"\"\"\n",
    "    \n",
    "    gene1 = zscores[zscores.index==pair[0]]\n",
    "    gene2 = zscores[zscores.index==pair[1]]\n",
    "    if gene1.shape[0] > 0 and gene2.shape[0] > 0:\n",
    "        gene1 = gene1.squeeze()\n",
    "        gene2 = gene2.squeeze()\n",
    "        return gene1.corr(gene2)\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dae5d30-be32-4b28-9033-17b7dd4b56ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def correlation_df(tc_zscores, gene_groups=syntelogs):\n",
    "    \"\"\"\n",
    "    Generate a pandas df containing the pairwise correlations of \n",
    "    expression patterns of genes as indicated in gene_groups\n",
    "\n",
    "    INPUTS:\n",
    "        tc_zscores:  a df containing the z-scored expression patterns of all genes, indexed by \"geneID\"\n",
    "        gene_groups: a df containing one column of \"geneIDs\" and one of groupings (\"Syntelog\")\n",
    "\n",
    "    OUTPUT:\n",
    "        a pandas df in the format of gene1 gene2 corr\n",
    "    \"\"\"\n",
    "    \n",
    "    synt_list = []\n",
    "    hap_a = []\n",
    "    hap_b = []\n",
    "    corrs = []\n",
    "    i = 0 # for status updates\n",
    "    n = len(gene_groups.Syntelog.unique())\n",
    "\n",
    "    #iterate over all the groupings\n",
    "    for synt in gene_groups.Syntelog.unique():\n",
    "\n",
    "        # update status\n",
    "        print(f\"Correlating Syntelog {i} of {n}\", end=\"\\r\")\n",
    "        i += 1\n",
    "        \n",
    "        #filter to the current syntelog group and get the list of geneIDs\n",
    "        cur_synt = gene_groups[gene_groups.Syntelog==synt]\n",
    "        cur_haps = cur_synt.geneID.unique()\n",
    "\n",
    "        # use combinations to get all pairwise combinations, regardless of orientation\n",
    "        for cur_pair in combinations(cur_haps,2):\n",
    "            cur_genes = cur_synt[cur_synt.geneID.isin(cur_pair)].geneID.to_list()\n",
    "            # calculate the correlation of that pare\n",
    "            cur_corr = calculate_pair_corr(tc_zscores, cur_genes)\n",
    "            #we only want to record the correlation is they're both present in the file\n",
    "            if cur_corr is not None:\n",
    "                synt_list.append(synt)\n",
    "                hap_a.append(cur_pair[0])\n",
    "                hap_b.append(cur_pair[1])\n",
    "                corrs.append(cur_corr)\n",
    "                \n",
    "    # convert all the output list to a dataframe\n",
    "    output = pd.DataFrame({\"Syntelog\":synt_list, \"Haplotype_1\":hap_a,\n",
    "                           \"Haplotype_2\":hap_b, \"Correlation\":corrs})\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0314ebd9-526e-4387-b2ea-390c7dc2244b",
   "metadata": {},
   "outputs": [],
   "source": [
    "broken = False\n",
    "time_courses = [\"shd\", \"lgd\", \"tuber\", \"tissue_rlog_tissue\", \"tissue_rlog_stress\"]\n",
    "mixer.music.load(attention_song)\n",
    "done = False\n",
    "for i, tc in enumerate(time_courses):\n",
    "    # figure out which files and types and columns to use based on which tc is current\n",
    "    print(f\"On {tc}\")\n",
    "    use_cols = None\n",
    "    if \"tissue\" in tc:\n",
    "        if \"stress\" in tc:\n",
    "            use_cols = \"stress\"\n",
    "        elif \"tissue\" in tc:\n",
    "            use_cols = \"tissue\"\n",
    "        use_tc = \"tissue\"\n",
    "        dtype = \"tissue\"\n",
    "    else:\n",
    "        use_tc = tc\n",
    "        dtype = \"timecourse\"\n",
    "\n",
    "    # make the z-score df\n",
    "    print(\"Calculating z-scores\", end=\"\\r\")\n",
    "    tc_zscores = avg_zscore_format(tc_dict[use_tc], dtype, use_cols)\n",
    "    \n",
    "    #make the correlation df and save it to a csv\n",
    "    output = correlation_df(tc_zscores)\n",
    "    output.to_csv(f\"{outfolder}{tc}_pairwise_corrs.csv\", index=False)\n",
    "    disp.clear_output(wait=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
